New bans + some other stuff

QuiCM opened the pull request:
* Bans completely overhauled
   - Bans are based on single pieces of identifiable info (e.g., one of: uuid, IP, name, account name) rather than an amalgamation of the lot
     - A generic user ban (account, IP, UUID) would result in three unique bans
   - Bans expire, but remain in the system. New bans for a given identifier can only be added once an existing ban for that identifier expires. Expired bans can still be inspected, but are  not actively displayed
     - I.E., you cannot have two enforced bans for the same UUID. Note that this is enforced in code, not in database.
  - Conversion path exists from old bans -> new bans

* Console automagically detects Terraria colour codes and attempts to display 'nearby' colours (only 'nearby' because ConsoleColor is pretty sparse, and Windows Cmd prompt doesn't properly support ANSI color codes yet)

* Shutdown now disconnects rather than kicks. There's something wrong with the kick code that can cause a connection drop rather than a clean disconnect

* Sudo should now properly forward `"quoted parameters"`

QuiCM committed:
Ban rewrite

QuiCM committed:
Merge branch 'general-devel' into new-bans

QuiCM committed:
Ban rewrite and various other adjustments

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
+  * All old ban routes in REST are now redirected. Please use `/v3/bans/*` for REST-based ban management.
+  * TShock recognizes and acts upon 4 main identifiers: UUID, IP, Player Name, Account name. This can be extended by plugins.
+  * By default, bans are no longer removed upon expiry or 'deletion'. Instead, they remain in the system. A new ban for an indentifier can be added once an existing ban has expired.
+* Server Console now understands Terraria color codes (e.g., `[c/FF00FF:Words]`) and prints the colored text to the console. Note that console colors are limited and thus only approximations. (@QuiCM)
```
Commit level: troll

QuiCM replied:
> 
> 
> This is a really really good improvement. I left comments asking questions here & there where I got confused. You don't need to address things really -- I don't think I see any show stoppers. I do see some potential confusion between using `identifier` and a `database ID that auto increments and is the primary key` though. I think it's worth renaming the DB column to a `ticket number` or something, to be less confusing to people. Users can specify `identifiers` which are criteria to identify someone who is banned, but the primary user interface requires users to pass around `database IDs`, and I can see this easily confusing someone.

This is a good point. The original plan was not to have an auto-inc column and for everything to be done around the identifiers. However this encountered an issue when I found that GUIDs are 128-bit integers in .NET, and I couldn't reasonably expect people to type out `ban del uuid:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...etc`.
I like the idea of 'ticket number' as opposed to an arbitrary 'unique ID'. Will adjust the naming with that in mind

> 
> I think we should consider just throwing away the concept of being "immune to ban." In practice, you can ban anyone with an IP. It's probably better to advise users not to give ban permissions to untrusted people, or restrict who can ban by what (maybe only some people can ban using an exact target, and that's a separate permission?). 

This is fair enough. I had this thought too when I re-added the ban immunity, because it technically only works if you ban on a player target. You could very easily do `ban add acc:hakusaro -e` and it wouldn't check if account 'hakusaro' was immune to bans.

>
>Finally, we don't consider other fancy things like wildcard banning or CIDR range banning, which might also be worth considering at some point.

I agree - I didn't go too in depth with expanding on identifiers we already exist. However, because of the flexibility of arbitrary identifiers, plugins should easily be able to add, validate, and enforce their own identifiers - all using the same database & structure as the vanilla TShock system. Theoretically it shouldn't be too difficult to add something like `cidr:127.0.0.1/24` as a valid identifier and check it on user join. I also attempted to design around TShock not enforcing anything if other plugins get in first. I may implement an example extension plugin just to see how simple or complex it is to extend the system

hakusaro replied:
Yeah -- in general, immunity is something tba. If anything, we should warn people that it's not foolproof.

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
   * If there is no section called "Upcoming changes" below this line, please add one with `## Upcoming changes` as the first line, and then a bulleted item directly after with the first change.
 
 ## Upcoming Changes
+* Overhauled Bans system. Bans are now based on 'identifiers'.(@QuiCM)
```
```suggestion
* Overhauled Bans system. Bans are now based on 'identifiers'. (@QuiCM)
```

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
 
 ## Upcoming Changes
+* Overhauled Bans system. Bans are now based on 'identifiers'.(@QuiCM)
+  * The old Bans table (`Bans`) has been deprecated. New bans will go in `PlayerBans`. Use `/ban convert` to convert to the new system. 
```
If it's non-destructive, you should just auto-upgrade users.

Given the importance of the system (and given that I'm reading the changelog first), you should also consider doing something like https://github.com/github/scientist where you raise at least some flag somewhere if a user who's joining would be banned using the new system and not the old system, and vice versa. Not at all required though.

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
 ## Upcoming Changes
+* Overhauled Bans system. Bans are now based on 'identifiers'.(@QuiCM)
+  * The old Bans table (`Bans`) has been deprecated. New bans will go in `PlayerBans`. Use `/ban convert` to convert to the new system. 
+  * All old ban routes in REST are now redirected. Please use `/v3/bans/*` for REST-based ban management.
```
Seems good.

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
+  * The old Bans table (`Bans`) has been deprecated. New bans will go in `PlayerBans`. Use `/ban convert` to convert to the new system. 
+  * All old ban routes in REST are now redirected. Please use `/v3/bans/*` for REST-based ban management.
+  * TShock recognizes and acts upon 4 main identifiers: UUID, IP, Player Name, Account name. This can be extended by plugins.
+  * By default, bans are no longer removed upon expiry or 'deletion'. Instead, they remain in the system. A new ban for an indentifier can be added once an existing ban has expired.
```
Perfect.

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
+  * TShock recognizes and acts upon 4 main identifiers: UUID, IP, Player Name, Account name. This can be extended by plugins.
+  * By default, bans are no longer removed upon expiry or 'deletion'. Instead, they remain in the system. A new ban for an indentifier can be added once an existing ban has expired.
+* Server Console now understands Terraria color codes (e.g., `[c/FF00FF:Words]`) and prints the colored text to the console. Note that console colors are limited and thus only approximations. (@QuiCM)
+* Fixed a bug in `/sudo` that prevented quoted arguments being forwarded properly. Example: `/sudo /user group "user name" "user group"` should now work correctly. (@QuiCM)
```
Awesome.

hakusaro commented on CHANGELOG.md:
```diff
@@ -13,6 +13,14 @@ This is the rolling changelog for TShock for Terraria. Use past tense when addin
+  * By default, bans are no longer removed upon expiry or 'deletion'. Instead, they remain in the system. A new ban for an indentifier can be added once an existing ban has expired.
+* Server Console now understands Terraria color codes (e.g., `[c/FF00FF:Words]`) and prints the colored text to the console. Note that console colors are limited and thus only approximations. (@QuiCM)
+* Fixed a bug in `/sudo` that prevented quoted arguments being forwarded properly. Example: `/sudo /user group "user name" "user group"` should now work correctly. (@QuiCM)
+* Shutting down the server should now correctly display the shutdown message to players rather than 'Lost connection'. (@QuiCM)
```
Superb.

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-						// Effective ban target assignment
-						List<TSPlayer> players = TSPlayer.FindByNameOrID(args.Parameters[1]);
+				args.Player.SendMessage("TShock Ban Help", Color.White);
+				args.Player.SendMessage("Available Ban commands:", Color.White);
```
I suggest adding at least one example of the most common type of ban here. When users are reading this, there's a good chance they're trying to ban someone _right now_, and now is not the ideal time for them to read more help pages. A simple example, e.g., `/ban add hackerperson hacking` or similar would suffice.

QuiCM replied:
Added a 'quick usage' line: `Quick usage: ban add {player's name} "Griefing"`

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-			string subcmd = args.Parameters.Count == 0 ? "help" : args.Parameters[0].ToLower();
-			switch (subcmd)
+			//Ban syntax:
+			// ban add <target> [reason] [duration] [flags (default: -a -u -ip)]
```
What're the units for duration?

QuiCM replied:
Added a comment about durations

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+					case "add":
+						args.Player.SendMessage("", Color.White);
+						args.Player.SendMessage("Ban Add Syntax", Color.White);
+						args.Player.SendMessage($"{"ban add".Color(Utils.BoldHighlight)} <{"Target".Color(Utils.RedHighlight)}> [{"Reason".Color(Utils.BoldHighlight)}] [{"Duration".Color(Utils.PinkHighlight)}] [{"Flags".Color(Utils.GreenHighlight)}]", Color.White);
```
I really wish like, idk, I hate horizontal scrolling

QuiCM replied:
Me too. It's vaguely easier to read if I use raw colour codes (i.e., `"[c/FFAAAA:ban add] stuff"` but far more difficult to reuse. Splitting lines doesn't add much readability either :(

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				switch (cmd)
+				{
+					case "add":
+						args.Player.SendMessage("", Color.White);
```
Interesting -- why?

QuiCM replied:
Was to split up messages a bit. Despite the snazzy colours, if you use a bunch of text-heavy commands in a row (eg `ban help`, `ban help add`) you end up with a big wall of text

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+						args.Player.SendMessage("Ban Add Syntax", Color.White);
+						args.Player.SendMessage($"{"ban add".Color(Utils.BoldHighlight)} <{"Target".Color(Utils.RedHighlight)}> [{"Reason".Color(Utils.BoldHighlight)}] [{"Duration".Color(Utils.PinkHighlight)}] [{"Flags".Color(Utils.GreenHighlight)}]", Color.White);
+						args.Player.SendMessage($"- {"Duration".Color(Utils.PinkHighlight)}: uses the format {"0d0m0s".Color(Utils.PinkHighlight)} to determine the length of the ban.", Color.White);
+						args.Player.SendMessage($"   Eg a value of {"10d30m0s".Color(Utils.PinkHighlight)} would represent 10 days, 30 minutes, 0 seconds.", Color.White);
```
What about a permaban? How long is that?

QuiCM replied:
Added a line about permanent banning

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				args.Player.SendMessage($"{"Banned by:".Color(Utils.BoldHighlight)} {ban.BanningUser.Color(Utils.GreenHighlight)} on {ban.BanDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({ban.GetPrettyTimeSinceBanString().Color(Utils.YellowHighlight)} ago)", Color.White);
+				if (ban.ExpirationDateTime < DateTime.UtcNow)
+				{
+					args.Player.SendMessage($"{"Banned expired:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({ban.GetPrettyExpirationString().Color(Utils.YellowHighlight)} ago)", Color.White);
```
```suggestion
					args.Player.SendMessage($"{"Ban expired:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({ban.GetPrettyExpirationString().Color(Utils.YellowHighlight)} ago)", Color.White);
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-								return;
-							}
-						}
+					args.Player.SendMessage($"{"Banned expires:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({remaining})", Color.White);
```
```suggestion
					args.Player.SendMessage($"{"Ban expires:".Color(Utils.BoldHighlight)} {ban.ExpirationDateTime.ToString("yyyy/MM/dd").Color(Utils.RedHighlight)} ({remaining})", Color.White);
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				for (int i = 2; i < args.Parameters.Count; i++)
+				{
+					var param = args.Parameters[i];
+					if (param != "-e" && param != "-a" && param != "-u" && param != "-n" && param != "-ip")
```
Could store this in an array so you can just check to see if param is in the array or not, rather than copypasta twice.

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-							targetGeneralizedName = target.Name;
-							success = TShock.Bans.AddBan(target.IP, target.Name, target.UUID, target.Account?.Name ?? "", banReason, false, args.Player.Account.Name,
-								banLengthInSeconds == 0 ? "" : DateTime.UtcNow.AddSeconds(banLengthInSeconds).ToString("s"));
+				//This is hacky, but because the flag values can go at any parameter, this is forcing the ordering of 'reason' and 'duration'
```
So the purpose here is to allow the flags to go in any order, and then make duration and reason order dependent? If so, you should like, reword to something that emphasizes the _goal_.

```suggestion
				//This is hacky, because we want the flag values to be at any position, but this forces us to order reason and duration consecutively
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-							success = TShock.Bans.AddBan(target.IP, target.Name, target.UUID, target.Account?.Name ?? "", banReason, false, args.Player.Account.Name,
-								banLengthInSeconds == 0 ? "" : DateTime.UtcNow.AddSeconds(banLengthInSeconds).ToString("s"));
+				//This is hacky, but because the flag values can go at any parameter, this is forcing the ordering of 'reason' and 'duration'
+				//while still allowing them to be arbitrarily placed in the parameter list
```
```suggestion
				
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				}
 
-						}
+				if (exactTarget)
```
I believe at this stage, if we have an exact target, we can be in a situation where no other flags were provided, in which case the defaults weren't set.

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+			//Ban syntax:
+			// ban add <target> [reason] [duration] [flags (default: -a -u -ip)]
+			//						Valid flags: -a (ban account name), -u (ban UUID), -n (ban character name), -ip (ban IP address), -e (exact, ban the identifier provided as 'target')
+			//						Unless -e is passed to the command, <target> is assumed to be a player or player index.
```
If -e is passed, how do you specify a target?

QuiCM replied:
If `-e` is passed, the `target` parameter is treated directly as an identifier rather than a player.
E.g., `ban add ip:127.0,0,1 -e` would add a ban for the identifier `ip:127.0.0.1`.
If we are not using `-e`, the `target` parameter is used to attempt to identify an online player

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
 
-						return;
+				string banReason = null;
+				void DoBan(string ident)
```
`Banish`? lol

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-							args.Player.SendErrorMessage("Invalid syntax! Proper syntax: {0}ban del <player>", Specifier);
-							return;
-						}
+						args.Player.SendWarningMessage($"Ban skipped for identifier: {ident}");
```
```suggestion
						args.Player.SendWarningMessage($"Ban failed for identifier: {ident}");
```

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-						}
+				if (exactTarget)
+				{
+					banResult = TShock.Bans.InsertBan(target, reason ?? "Banned", args.Player.Account.Name, DateTime.UtcNow, expiration);
```
I haven't yet gotten to `InsertBan`, but am I correct in assuming that if a player is banned using an `exact` ban while online, they won't be booted?

hakusaro replied:
Does this duplicate `DoBan`?

QuiCM replied:
>
>I haven't yet gotten to InsertBan, but am I correct in assuming that if a player is banned using an exact ban while online, they won't be booted?

Yep

>
>Does this duplicate `DoBan`?

Also yep

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+					DoBan($"{Identifiers.IP}{player.IP}");
+				}
+
+				//Using the ban reason to determine if a ban actually happened or not is messy, but it works
```
o.o

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+				//Using the ban reason to determine if a ban actually happened or not is messy, but it works
+				if (banReason != null)
+				{
+					player.Disconnect($"You have been banned: {banReason}.");
```
ForceKick? Does that even do anything different?

QuiCM replied:
Kick causes the 'Lost Connection' screen with worrying frequency at the moment, while Disconnect triggers a clean disconnect displaying the disconnection reason

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
-						};
+				if (TShock.Bans.RemoveBan(banId))
+				{
+					args.Player.SendSuccessMessage($"Ban {banId} has now been marked as expired.");
```
Could be something we'd want to log somewhere (e.g., "who removed this ban?")

hakusaro commented on TShockAPI/Commands.cs:
```diff
@@ -1267,328 +1267,368 @@ private static void Kick(CommandArgs args)
+						args.Player.SendMessage("", Color.White);
+						args.Player.SendMessage("Ban List Syntax", Color.White);
+						args.Player.SendMessage($"{"ban list".Color(Utils.BoldHighlight)} [{"Page".Color(Utils.PinkHighlight)}]", Color.White);
+						args.Player.SendMessage("- Lists active bans. Color trends towards green as the ban approaches expiration", Color.White);
```
What about the inactive bans?

QuiCM replied:
Inactive bans are currently invisible unless you know the ticket number. Could add a parameter to include those in the list

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -39,15 +70,13 @@ public BanManager(IDbConnection db)
-									new SqlColumn("UUID", MySqlDbType.Text),
+			var table = new SqlTable("PlayerBans",
+									new SqlColumn("Id", MySqlDbType.Int32) { Primary = true, AutoIncrement = true },
+									new SqlColumn("Identifier", MySqlDbType.Text),
```
Perhaps call this something other than `Identifier` because `Id` can be thought of as shortened "identifier." On `	args.Player.SendErrorMessage("No ban found matching the given identifier");`, `identifier` was used to refer to the `Id` column, too.

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -62,231 +91,319 @@ public BanManager(IDbConnection db)
-				TShock.Log.Error(ex.ToString());
-				Console.WriteLine(ex.StackTrace);
+				//We consider a ban to be valid if the start time is before now and the end time is after now, and the player is not immune to bans
+				args.Valid = (DateTime.UtcNow > args.Ban.BanDateTime && DateTime.UtcNow < args.Ban.ExpirationDateTime) && !args.Player.HasPermission(Permissions.immunetoban);
```
Can you use an exact target to ban an immune player?

QuiCM replied:
Yep

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -62,231 +91,319 @@ public BanManager(IDbConnection db)
+
+			if (!args.Valid)
+			{
+				string message = $"Ban was invalidated: {(args.Message ?? "no further information provided.")}";
```
```suggestion
				string message = $"Ban was not valid: {(args.Message ?? "no further information provided.")}";
```

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -62,231 +91,319 @@ public BanManager(IDbConnection db)
+		/// <param name="identifier">Identifier to search with</param>
+		/// <param name="currentOnly">Whether or not to exclude expired bans</param>
+		/// <returns></returns>
+		public IEnumerable<Ban> RetrieveBansByIdentifier(string identifier, bool currentOnly = true)
```
What if I pass `Id`? That screams that it's a valid identifier to me...

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -62,231 +91,319 @@ public BanManager(IDbConnection db)
+				{
+					var uniqueId = reader.Get<int>("Id");
+					var ident = reader.Get<string>("Identifier");
+					var id = reader.Get<string>("Identifier");
```
Both `id` and `ident` are set to the same thing here.

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -330,122 +441,129 @@ public enum BanSortMethod
-				return _method == BanSortMethod.AddedNewestToOldest || _method == BanSortMethod.ExpirationSoonestToLatest ? 1 : -1;
-			}
+	/// <summary>
+	/// Event args used for formalized bans
```
Formalized?

QuiCM replied:
Changed to 'completed'. Ban objects used in these eventargs are complete and have a valid ticket number

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -330,122 +441,129 @@ public enum BanSortMethod
-		/// <value>The IP Address</value>
-		public string IP { get; set; }
-
+		public const string IP = "ip:";
```
Okay, I officially think these should be in the help menu somewhere.

hakusaro commented on TShockAPI/DB/BanManager.cs:
```diff
@@ -330,122 +441,129 @@ public enum BanSortMethod
 		/// </summary>
-		/// <value>The UUID</value>
-		public string UUID { get; set; }
+		public int UniqueId { get; set; }
```
Maybe call this a ticket number?

hakusaro commented on TShockAPI/Extensions/DbExt.cs:
```diff
@@ -156,10 +185,10 @@ public static SqlType GetSqlType(this IDbConnection conn)
 				(s, i) => s.IsDBNull(i) ? null : (object)s.GetInt32(i)
 			},
-			{
+			/*{
```
?

QuiCM replied:
This is an interesting one. Our defined override for Int64 was throwing exceptions. After extending `Get<T>` to hypothetically work on anything as long as you pass the right type, problem went away. I may need to retest this

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3592,7 +3592,7 @@ private static bool HandleSyncExtraValue(GetDataHandlerArgs args)
 			}
 
-			return false;
+			return true;
```
Wait, why?

QuiCM replied:
Oops, that wasn't meant to be in this particular PR

hakusaro reviewed (commented):
This is a really really good improvement. I left comments asking questions here & there where I got confused. You don't need to address things really -- I don't think I see any show stoppers. I do see some potential confusion between using `identifier` and a `database ID that auto increments and is the primary key` though. I think it's worth renaming the DB column to a `ticket number` or something, to be less confusing to people. Users can specify `identifiers` which are criteria to identify someone who is banned, but the primary user interface requires users to pass around `database IDs`, and I can see this easily confusing someone.

I think we should consider just throwing away the concept of being "immune to ban." In practice, you can ban anyone with an IP. It's probably better to advise users not to give ban permissions to untrusted people, or restrict who can ban by what (maybe only some people can ban using an exact target, and that's a separate permission?). Finally, we don't consider other fancy things like wildcard banning or CIDR range banning, which might also be worth considering at some point.

Overall though, I'd say that any love given to this system is good. Thank you so much for working on it!!

QuiCM committed:
Refactor 'Id' to 'TicketNumber' and extend the identifier model

QuiCM committed:
Remove ban immunity at the command level.

Ban immunity will be managed when a ban is validated

QuiCM committed:
Rename unique ID -> ticket number
Also get ban conversion to execute cleanly and clean up after itself

QuiCM committed:
Some more type checking on QueryScalar

QuiCM committed:
Remove ban conversion command as it now happens automatically

QuiCM committed:
Refactoring based on suggestions

QuiCM committed:
Refactored for 'ticket number' naming
Added 'ban help identifiers' to list known identifiers
Fixed some typos and restructured some logic

QuiCM committed:
Revert syncextravalue change

QuiCM committed:
Offload ban checking to BanManager
Remove ban immunity

QuiCM committed:
Obsolete the Permissions.immunetoban field

QuiCM committed:
Re-enable Int64 override for IDataReader.Get

hakusaro reviewed (approved):
![lgtm_6](https://user-images.githubusercontent.com/532647/100960886-0ff08a00-34d6-11eb-8313-ffb1f486eda2.gif)

In general, I think the changes you've made will help a lot! Everything looks good to me, though I didn't read it with a super fine tooth comb. I do really think the help for identifiers is top tier though, and shows that it's a good design. I honestly think this is a wicked improvement over what we've had before.

I totally understand how you got here, with respect to whether or not to use an autoincrementing ID. It makes total sense now. But yeah, all good now 👍

Tysm for working on this!! 💚
